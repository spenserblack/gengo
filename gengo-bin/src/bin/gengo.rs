use std::process::exit;
fn main() {
    let cli = gengo_bin::cli::new();
    if let Err(err) = cli.run(std::io::stdout(), std::io::stderr()) {
        eprintln!("{}", err);
        exit(1);
    }
}
