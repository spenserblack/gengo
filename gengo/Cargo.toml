[package]
name = "gengo"
description = { workspace = true }
readme = "README.md"
version = { workspace = true }
edition = { workspace = true }
repository = { workspace = true }
license = { workspace = true }

keywords = { workspace = true }

[dependencies]
git2 = { version = "0.17", default-features = false }
glob = "0.3"
indexmap = { version = "2", features = ["serde"] }
once_cell = "1"
regex = "1"
serde = { version = "1", features = ["derive"] }
serde_json = "1"
serde_yaml = "0.9"

[build-dependencies]
indexmap = { version = "2", features = ["serde"] }
serde_json = "1"
serde_yaml = "0.9"

[dev-dependencies]
criterion = "0.5"
insta = "1"
rstest = "0.18"

[[bench]]
# NOTE: We run on this repo as a real-world example
name = "run_on_self"
harness = false
