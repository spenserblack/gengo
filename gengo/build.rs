use indexmap::IndexMap;
use proc_macro2::{Ident, Literal, Span};
use quote::quote;
use std::collections::HashMap;
use std::env;
use std::error::Error;
use std::fs;
use std::path::Path;

const LANGUAGES: &str = include_str!("./languages.yaml");

const MINIMUM_PRIORITY: u8 = 0;
const MAXIMUM_PRIORITY: u8 = 100;
const DEFAULT_PRIORITY: u8 = 50;

/// Converts `languages.yaml` to minified JSON and writes it to
/// `languages.json`.
fn main() -> Result<(), Box<dyn Error>> {
    // TODO This looks messy, and can use cleanup.
    let languages: IndexMap<String, serde_json::Value> = serde_yaml::from_str(LANGUAGES)?;
    let languages_target_dir = Path::new(&env::var("OUT_DIR")?).join("languages");
    fs::create_dir_all(&languages_target_dir)?;

    struct LanguageDefinition {
        variant: Ident,
        /// See `Category`.
        category: Ident,
        name: Literal,
        color: Literal,
        priority: Literal,
        extensions: Vec<String>,
        filenames: Vec<String>,
        interpreters: Vec<String>,
        patterns: Vec<String>,
        heuristics: Vec<String>,
    }

    let language_definitions: Vec<_> = languages
        .iter()
        .map(|(language_name, language_attrs)| {
            let language_attrs = language_attrs
                .as_object()
                .expect("language attributes to be an object");
            let variant = rustify_language_name(language_name);
            let variant = Ident::new(&variant, Span::call_site());

            let category = language_attrs["category"]
                .as_str()
                .expect("category to be a string");
            let category = match category {
                "data" => "Data",
                "markup" => "Markup",
                "programming" => "Programming",
                "prose" => "Prose",
                "query" => "Query",
                unknown => unreachable!("Category {unknown}"),
            };
            let category = Ident::new(category, Span::call_site());

            let name = Literal::string(language_name);

            let color = language_attrs["color"]
                .as_str()
                .expect("color to be a string");
            let color = Literal::string(color);

            let priority = language_attrs
                .get("priority")
                .map(|priority| {
                    let priority = priority.as_u64().expect("priority to be a number");
                    assert!(
                        priority >= MINIMUM_PRIORITY.into() && priority <= MAXIMUM_PRIORITY.into(),
                        "priority to be between {MINIMUM_PRIORITY} and {MAXIMUM_PRIORITY}"
                    );
                    priority as u8
                })
                .unwrap_or(DEFAULT_PRIORITY);
            let priority = Literal::u8_unsuffixed(priority);

            let matchers = language_attrs["matchers"]
                .as_object()
                .expect("matchers to be an object");

            let extensions = matchers
                .get("extensions")
                .map(|extensions| {
                    extensions
                        .as_array()
                        .expect("extensions to be an array")
                        .to_owned()
                })
                .unwrap_or_default()
                .iter()
                .map(|extension| {
                    extension
                        .as_str()
                        .expect("extension to be a string")
                        .to_string()
                })
                .collect();

            let filenames = matchers
                .get("filenames")
                .map(|filenames| {
                    filenames
                        .as_array()
                        .expect("filenames to be an array")
                        .to_owned()
                })
                .unwrap_or_default()
                .iter()
                .map(|filename| {
                    filename
                        .as_str()
                        .expect("filename to be a string")
                        .to_string()
                })
                .collect();

            let interpreters = matchers
                .get("interpreters")
                .map(|interpreters| {
                    interpreters
                        .as_array()
                        .expect("interpreters to be an array")
                        .to_owned()
                })
                .unwrap_or_default()
                .iter()
                .map(|interpreter| {
                    interpreter
                        .as_str()
                        .expect("interpreter to be a string")
                        .to_string()
                })
                .collect();

            let patterns = matchers
                .get("patterns")
                .map(|patterns| {
                    patterns
                        .as_array()
                        .expect("patterns to be an array")
                        .to_owned()
                })
                .unwrap_or_default()
                .iter()
                .map(|pattern| {
                    pattern
                        .as_str()
                        .expect("pattern to be a string")
                        .to_string()
                })
                .collect();

            let heuristics = language_attrs
                .get("heuristics")
                .map(|heuristics| {
                    heuristics
                        .as_array()
                        .expect("heuristics to be an array")
                        .to_owned()
                        .iter()
                        .map(|heuristic| {
                            heuristic
                                .as_str()
                                .expect("heuristic to be a string")
                                .to_string()
                        })
                        .collect()
                })
                .unwrap_or_default();

            LanguageDefinition {
                variant,
                category,
                name,
                color,
                priority,
                extensions,
                filenames,
                interpreters,
                patterns,
                heuristics,
            }
        })
        .collect();

    let variants = language_definitions.iter().map(|def| &def.variant);
    let language = quote! {
        /// The type of language. Returned by language detection.
        #[non_exhaustive]
        #[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
        pub enum Language {
            #(#variants,)*
        }
    };
    fs::write(
        languages_target_dir.join("language.rs"),
        language.to_string(),
    )?;

    let category_mappings = language_definitions.iter().map(
        |LanguageDefinition {
             variant, category, ..
         }| {
            quote! {
                Self::#variant => Category::#category
            }
        },
    );
    let category_mixin = quote! {
        impl Language {
            /// Gets the category of the language.
            pub const fn category(&self) -> Category {
                match self {
                    #(#category_mappings ,)*
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("category_mixin.rs"),
        category_mixin.to_string(),
    )?;

    let name_mappings =
        language_definitions
            .iter()
            .map(|LanguageDefinition { variant, name, .. }| {
                quote! {
                    Self::#variant => #name
                }
            });
    let name_mixin = quote! {
        impl Language {
            /// Gets the name of the language.
            pub const fn name(&self) -> &'static str {
                match self {
                    #(#name_mappings ,)*
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("name_mixin.rs"),
        name_mixin.to_string(),
    )?;

    let reverse_variant_mappings =
        language_definitions
            .iter()
            .map(|LanguageDefinition { variant, .. }| {
                let variant_name = variant.to_string();
                quote! {
                    #variant_name => Some(Self::#variant)
                }
            });
    let parse_variant_mixin = quote! {
        impl Language {
            /// Converts a variant's name back to the language.
            fn parse_variant(name: &str) -> Option<Self> {
                match name {
                    #(#reverse_variant_mappings ,)*
                    _ => None,
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("parse_variant_mixin.rs"),
        parse_variant_mixin.to_string(),
    )?;

    let color_mappings =
        language_definitions
            .iter()
            .map(|LanguageDefinition { variant, color, .. }| {
                quote! {
                    Self::#variant => #color
                }
            });
    let color_mixin = quote! {
        impl Language {
            /// Gets the color associated with the language.
            pub const fn color(&self) -> &'static str {
                match self {
                    #(#color_mappings ,)*
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("color_mixin.rs"),
        color_mixin.to_string(),
    )?;

    let priority_mappings = language_definitions.iter().map(
        |LanguageDefinition {
             variant, priority, ..
         }| {
            quote! {
                Self::#variant => #priority
            }
        },
    );
    let priority_mixin = quote! {
        impl Language {
            /// Gets the priority of the language. This is useful for sorting languages
            /// when multiple languages are detected.
            pub const fn priority(&self) -> u8 {
                match self {
                    #(#priority_mappings ,)*
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("priority_mixin.rs"),
        priority_mixin.to_string(),
    )?;

    let extension_to_langs: HashMap<_, Vec<_>> = language_definitions.iter().fold(
        HashMap::new(),
        |map,
         LanguageDefinition {
             variant,
             extensions,
             ..
         }| {
            extensions.iter().fold(map, |mut map, extension| {
                map.entry(extension.clone())
                    .or_default()
                    .push(variant.clone());
                map
            })
        },
    );
    let extension_to_langs_mappings = extension_to_langs.iter().map(|(extension, langs)| {
        quote! {
            #extension => vec![#(Self::#langs),*]
        }
    });
    let from_extension_mixin = quote! {
        impl Language {
            /// Gets languages by extension.
            pub fn from_extension(extension: &str) -> Vec<Self> {
                match extension {
                    #(#extension_to_langs_mappings ,)*
                    _ => vec![],
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("from_extension_mixin.rs"),
        from_extension_mixin.to_string(),
    )?;

    let filenames_to_langs: HashMap<_, Vec<_>> = language_definitions.iter().fold(
        HashMap::new(),
        |map,
         LanguageDefinition {
             variant, filenames, ..
         }| {
            filenames.iter().fold(map, |mut map, filename| {
                map.entry(filename.clone())
                    .or_default()
                    .push(variant.clone());
                map
            })
        },
    );
    let filenames_to_langs_mappings = filenames_to_langs.iter().map(|(filename, langs)| {
        quote! {
            #filename => vec![#(Self::#langs),*]
        }
    });
    let from_filename_mixin = quote! {
        impl Language {
            /// Gets languages by filename.
            pub fn from_filename(filename: &str) -> Vec<Self> {
                match filename {
                    #(#filenames_to_langs_mappings ,)*
                    _ => vec![],
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("from_filename_mixin.rs"),
        from_filename_mixin.to_string(),
    )?;

    let interpreters_to_langs: HashMap<_, Vec<_>> = language_definitions.iter().fold(
        HashMap::new(),
        |map,
         LanguageDefinition {
             variant,
             interpreters,
             ..
         }| {
            interpreters.iter().fold(map, |mut map, interpreter| {
                map.entry(interpreter.clone())
                    .or_default()
                    .push(variant.clone());
                map
            })
        },
    );
    let interpreter_to_langs_mappings = interpreters_to_langs.iter().map(|(interpreter, langs)| {
        quote! {
            #interpreter => vec![#(Self::#langs),*]
        }
    });
    let from_interpreter_mixin = quote! {
        impl Language {
            /// Gets languages by interpreter (typically found as part of a shebang).
            pub fn from_interpreter(interpreter: &str) -> Vec<Self> {
                match interpreter {
                    #(#interpreter_to_langs_mappings ,)*
                    _ => vec![],
                }
            }
        }
    };
    fs::write(
        languages_target_dir.join("from_interpreter_mixin.rs"),
        from_interpreter_mixin.to_string(),
    )?;

    let glob_matchers = language_definitions
        .iter()
        .filter(|def| !def.patterns.is_empty())
        .map(
            |LanguageDefinition {
                 variant, patterns, ..
             }| {
                quote! {
                   (
                       vec![#(#patterns),*],
                       Language::#variant,
                   )
                }
            },
        );
    let glob_mappings_mixin = quote! {
        impl Language {
            /// Gets the mappings used to map a glob to its language.
            fn glob_mappings() -> Vec<(Vec<&'static str>, Self)> {
                vec![#(#glob_matchers),*]
            }
        }
    };
    fs::write(
        languages_target_dir.join("glob_mappings_mixin.rs"),
        glob_mappings_mixin.to_string(),
    )?;

    let heuristic_tuples = language_definitions
        .iter()
        .filter(|language_definition| !language_definition.heuristics.is_empty())
        .map(
            |LanguageDefinition {
                 variant,
                 heuristics,
                 ..
             }| {
                quote! {
                    (Self::#variant, vec![#(#heuristics),*])
                }
            },
        );
    let heuristic_mappings_mixin = quote! {
        impl Language {
            /// Gets the heuristics used to determine a language.
            fn heuristic_mappings() -> Vec<(Self, Vec<&'static str>)> {
                vec![#(#heuristic_tuples ,)*]
            }
        }
    };
    fs::write(
        languages_target_dir.join("heuristic_mappings_mixin.rs"),
        heuristic_mappings_mixin.to_string(),
    )?;

    Ok(())
}

/// Converts a language name to a valid Rust identifier to be used as an enum
/// variant.
///
/// # Examples
///
/// - `".Env"` -> `"Dotenv"`
/// - `"C++"` -> `"CPlusPlus"`
/// - `"C#"` -> `"CSharp"`
/// - `"HTML"` -> `"Html"`
/// - `"JavaScript"` -> `"Javascript"`
/// - `"Batch File"` -> `"BatchFile"`
fn rustify_language_name(name: &str) -> String {
    let name = asciiify_language_name(name);

    // HACK This will break if there are any leading, trailing, or consecutive
    //      spaces in the name.
    let name = name.split(' ').fold(String::new(), |name, word| {
        let mut chars = word.chars();
        // NOTE If there is a special character like ß it will become SS, but
        //      that should never happen.
        let first_char = chars.next().unwrap().to_uppercase();
        assert!(first_char.len() == 1);
        let rest = chars
            .map(|c| c.to_lowercase().to_string())
            .collect::<String>();
        format!("{name}{first_char}{rest}")
    });
    name
}

/// Replaces special characters in a language name with their ASCII
/// equivalents.
fn asciiify_language_name(name: &str) -> String {
    // NOTE .Env is a special case because the special character is at the beginning
    //      and it should be one word.
    if name == ".Env" {
        return "Dotenv".to_string();
    }
    // NOTE Maps special characters to their ASCII equivalents.
    let mappings = [("-", ""), ("'", ""), ("+", "Plus"), ("#", "Sharp")];

    let name = mappings
        .iter()
        .fold(name.to_string(), |name, (pattern, replacement)| {
            // NOTE Adding a leading space to the replacement to ensure that it
            //      is treated as a word boundary.
            name.replace(pattern, &format!(" {replacement}"))
        });

    name
}
