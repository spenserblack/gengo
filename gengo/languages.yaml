# Definition of languages supported by Gengo
#
# # Format
#
# ```yaml
# # The name of the language.
# Language:
#
#   # data, markup, programming, prose, or query
#   category:
#
#   # The color associated with the language (unused).
#   color:
#
#   # Regexes unique to the contents of the file to resolve multiple matches. These will
#   # be used if the matchers (see below) cannot narrow down a file to a single
#   # language.
#   heuristics:
#     - "(?m)^macro_rules!"
#
#   # Entries to match against files. This is the first pass before more expensive
#   # strategies are used.
#   matchers:
#
#     # The file extensions associated with the language (optional)
#     extensions:
#       - "rs"
#
#     # The filenames associated with the language (optional)
#     filenames:
#       - "Dockerfile"
#
#     # Used for checking shebangs.
#     interpreters: # The interpreter(s) associated with the language (optional)
#       - "python"
#       - "python3"
#
#     # Glob patterns associated with the language (optional)
#     patterns:
#       # See the format of the `glob` crate: https://docs.rs/glob/latest/glob/struct.Pattern.html
#       - "__snapshots__/*.snap"
#
#   # Value between 0 and 100 to pick a language if multiple matches couldn't be
#   # narrowed down (optional, default: 50)
#   priority: 50
# ```
#
# WARNING: Please alphabetize!
---
ABAP:
  category: programming
  color: "#3C3C3C"
  matchers:
    extensions:
      - abap
ABNF:
  category: data
  color: "#ABFABF"
  matchers:
    extensions:
      - abnf
Ada:
  category: programming
  color: "#005A00"
  matchers:
    extensions:
      - ada
      - adb
      - ads
Assembly:
  category: programming
  color: "#33AA33"
  matchers:
    extensions:
      - asm
      - s
AutoHotkey:
  category: programming
  color: "#334455"
  matchers:
    extensions:
      - ahk
C:
  category: programming
  color: "#8888CC"
  matchers:
    extensions:
      - c
      - h
"C#":
  category: programming
  color: "#178600"
  heuristics:
    - "^\\s*(using\\s+[A-Z][\\s\\w.]+;|namespace\\s*[\\w\\.]+\\s*(\\{|;)|\\/\\/)"
  matchers:
    extensions:
      - cake
      - cs
      - csx
"C++":
  category: programming
  color: "#88CC88"
  matchers:
    extensions:
      - c++
      - cpp
      - h
      - h++
      - hh
      - hpp
CMake:
  category: programming
  color: "#CCCCCC"
  matchers:
    extensions:
      - cmake
    filenames:
      - CMakeLists.txt
CSS:
  category: markup
  color: "#AA88AA"
  matchers:
    extensions:
      - css
Ceylon:
  category: programming
  color: "#F7941E"
  matchers:
    extensions:
      - ceylon
Clojure:
  category: programming
  color: "#77F212"
  matchers:
    extensions:
      - clj
      - cljc
      - cljs
CoffeeScript:
  category: programming
  color: "#C0FFEE"
  matchers:
    extensions:
      - coffee
    interpreters:
      - coffee
ColdFusion:
  category: programming
  color: "#001C57"
  matchers:
    extensions:
      - cfm
Coq:
  category: programming
  color: "#D0B68C"
  matchers:
    extensions:
      - v
Crystal:
  category: programming
  color: "#000000"
  matchers:
    extensions:
      - cr
    interpreters:
      - crystal
D:
  category: programming
  color: "#B03931"
  matchers:
    extensions:
      - d
Dart:
  category: programming
  color: "#238BDA"
  matchers:
    extensions:
      - dart
Docker:
  category: programming
  color: "#2496ED"
  matchers:
    filenames:
      - "Dockerfile"
    patterns:
      - "Dockerfile.*" # Dockerfile.<extension>
Elixir:
  category: programming
  color: "#6B5674"
  matchers:
    extensions:
      - ex
      - exs
    interpreters:
      - elixir
Elm:
  category: programming
  color: "#1293D8"
  matchers:
    extensions:
      - elm
Emacs Lisp:
  category: programming
  color: "#7F5AB6"
  matchers:
    extensions:
      - el
Emojicode:
  category: programming
  color: "#FCEA2B"
  matchers:
    extensions:
      - emojic
      - "üçá"
Erlang:
  category: programming
  color: "#A90433"
  matchers:
    extensions:
      - erl
      - hrl
"F#":
  category: programming
  color: "#F8008F"
  matchers:
    extensions:
      - fs
      - fsx
GitHub Workflow:
  category: programming
  color: "#2088FF"
  matchers:
    patterns:
      - ".github/workflows/*.yaml"
      - ".github/workflows/*.yml"
  priority: 75
Go:
  category: programming
  color: "#00ADD8"
  matchers:
    extensions:
      - go
GraphQL:
  category: query
  color: "#E10098"
  matchers:
    extensions:
      - graphql
Groovy:
  category: programming
  color: "#4298B8"
  matchers:
    extensions:
      - groovy
    interpreters:
      - groovy
HTML:
  category: markup
  color: "#E96228"
  matchers:
    extensions:
      - html
Haskell:
  category: programming
  color: "#5E5086"
  matchers:
    extensions:
      - hs
Ignore List:
  category: data
  color: "#330000"
  matchers:
    filenames:
      - ".dockerignore"
      - ".gitignore"
      - ".npmignore"
      - ".vercelignore"
JSON:
  category: data
  color: "#AAAAAA"
  matchers:
    extensions:
      - json
    filenames:
      - .all-contributorsrc
  priority: 75
JSON with Comments:
  category: data
  color: "#CCCCCC"
  heuristics:
    - "(?m)^\\s*/[/\\*]"
  matchers:
    extensions:
      - json
      - jsonc
    filenames:
      - devcontainer.json
      - tsconfig.json
    patterns:
      - .vscode/*.json
      - tsconfig.*.json
  priority: 25
Java:
  category: programming
  color: "#5283A2"
  matchers:
    extensions:
      - java
JavaScript:
  category: programming
  color: "#F0DC4E"
  matchers:
    extensions:
      - js
      - jsx
    interpreters:
      - node
Julia:
  category: programming
  color: "#9558B2"
  matchers:
    extensions:
      - jl
Jupyter Notebook:
  category: markup
  color: "#F37726"
  matchers:
    extensions:
      - ipynb
Kotlin:
  category: programming
  color: "#7F52FF"
  matchers:
    extensions:
      - kt
      - kts
Lua:
  category: programming
  color: "#02027D"
  matchers:
    extensions:
      - lua
    interpreters:
      - lua
Makefile:
  category: programming
  color: "#6B482F" # Arbitrary brown color representing a Gnu
  matchers:
    filenames:
      - "Makefile"
    patterns:
      - "Makefile.*" # Makefile.<extension>
Markdown:
  category: prose
  color: "#03A7DD"
  matchers:
    extensions:
      - markdown
      - md
Mermaid:
  category: markup
  color: "#FF3670"
  matchers:
    extensions:
      - mermaid
      - mmd
Nim:
  category: programming
  color: "#ffe953"
  matchers:
    extensions:
      - nim
Nix:
  category: programming
  color: "#6898D3" # Average of the two colors in the logo
  matchers:
    extensions:
      - nix
PHP:
  category: programming
  color: "#7A86B8"
  matchers:
    extensions:
      - php
Perl:
  category: programming
  color: "#51547F"
  matchers:
    extensions:
      - cow # cowsay files are Perl
      - pl
    interpreters:
      - perl
Plain Text:
  category: prose
  color: "#000000"
  matchers:
    extensions:
      - text
      - txt
    filenames:
      - "LICENCE"
      - "LICENSE"
PowerShell:
  category: programming
  color: "#012456"
  matchers:
    extensions:
      - ps1
    interpreters:
      - pwsh
Pug:
  category: markup
  color: "#A86454"
  matchers:
    extensions:
      - pug
PureScript:
  category: programming
  color: "#1D222D"
  matchers:
    extensions:
      - purs
Python:
  category: programming
  color: "#3472A6"
  matchers:
    extensions:
      - py
    interpreters:
      - python
      - python2
      - python3
Python Requirements File:
  category: data
  color: "#FFD342"
  matchers:
    filenames:
      - "requirements.txt"
    patterns:
      - "requirements-*.txt"
  priority: 75
R:
  category: programming
  color: "#1F66B7" # Average of the two colors used in the logo gradient: https://www.r-project.org/logo/Rlogo.svg
  matchers:
    extensions:
      - R
Regex:
  category: data
  color: "#44E03F"
  matchers:
    extensions:
      - regex
Ruby:
  category: programming
  color: "#D21304"
  matchers:
    extensions:
      - gemspec
      - rb
    filenames:
      - Gemfile
      - Rakefile
    interpreters:
      - ruby
Rust:
  category: programming
  color: "#DD3515"
  matchers:
    extensions:
      - rs
SQL:
  category: query
  color: "#FFBF1E"
  matchers:
    extensions:
      - sql
Sass:
  category: markup
  color: "#CF649A"
  # NOTE: Sass has two syntaxes. See https://sass-lang.com/guide/
  matchers:
    extensions:
      - sass
      - scss
Shell:
  category: programming
  color: "#262E28"
  matchers:
    extensions:
      - bash
      - fish
      - sh
      - zsh
    interpreters:
      - bash
      - fish
      - sh
      - zsh
Solidity:
  category: programming
  color: "#2B247C"
  matchers:
    extensions:
      - sol
Svelte:
  category: programming
  color: "#FF3E00"
  matchers:
    extensions:
      - svelte
TOML:
  category: data
  color: "#9C4221"
  matchers:
    extensions:
      - toml
    filenames:
      - Cargo.lock
TypeScript:
  category: programming
  color: "#2F74C0"
  heuristics:
    - "(?m)^/// <reference "
    - "(?m)^export\\s+\\w[\\w\\d_]*?"
    - "(?m)^import\\s+"
  matchers:
    extensions:
      - ts
      - tsx
    interpreters:
      - deno
Vim Script:
  category: programming
  color: "#019833"
  matchers:
    extensions:
      - vim
    filenames:
      - ".vimrc"
Vue:
  category: programming
  color: "#3FB27F"
  matchers:
    extensions:
      - vue
XML:
  category: data
  color: "#005FAF"
  heuristics:
    - "<TS version=\"\\d+(?:\\.d+)+\" language=\""
  matchers:
    extensions:
      - ts
      - xml
  priority: 25
YAML:
  category: data
  color: "#CC1018"
  matchers:
    extensions:
      - yaml
      - yml
Zig:
  category: programming
  color: "#F7A41D"
  matchers:
    extensions:
      - zig
