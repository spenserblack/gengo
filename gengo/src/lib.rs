pub use builder::Builder;
use documentation::Documentation;
use generated::Generated;
use git2::{AttrCheckFlags, AttrValue, Blob, Commit, ObjectType, Repository, Tree};
use indexmap::IndexMap;
pub use languages::analyzer::Analyzers;
use languages::Category;
pub use languages::Language;
use std::error::Error;
use std::path::{Path, PathBuf};
use vendored::Vendored;

mod builder;
mod documentation;
mod generated;
pub mod languages;
mod vendored;

/// The main entry point for Gengo.
pub struct Gengo {
    repository: Repository,
    analyzers: Analyzers,
    read_limit: usize,
    documentation: Documentation,
    generated: Generated,
    vendored: Vendored,
}

// TODO parse .gitattributes to get language overrides.
impl Gengo {
    const ATTR_CHECK_FLAGS: [AttrCheckFlags; 2] =
        [AttrCheckFlags::NO_SYSTEM, AttrCheckFlags::INDEX_THEN_FILE];
    /// Resolves a revision to a commit.
    fn rev(&self, rev: &str) -> Result<Commit, Box<dyn Error>> {
        let reference = self.repository.revparse_single(rev)?;
        let commit = reference.peel_to_commit()?;
        Ok(commit)
    }

    /// Analyzes each file in the repository at the given revision.
    pub fn analyze(&self, rev: &str) -> Result<IndexMap<PathBuf, Entry>, Box<dyn Error>> {
        let mut results = IndexMap::new();
        let commit = self.rev(rev)?;
        let tree = commit.tree()?;
        self.analyze_tree("", &tree, &mut results)?;
        Ok(results)
    }

    fn analyze_tree(
        &self,
        root: &str,
        tree: &Tree,
        results: &mut IndexMap<PathBuf, Entry>,
    ) -> Result<(), Box<dyn Error>> {
        for entry in tree.iter() {
            let object = entry.to_object(&self.repository)?;
            match entry.kind() {
                Some(ObjectType::Tree) => {
                    let path = entry.name().ok_or("invalid path")?;
                    let tree = object.as_tree().expect("object to be a tree");
                    let path = Path::new(root).join(path);
                    let path = path.to_str().ok_or("invalid path")?;

                    self.analyze_tree(path, tree, results)?;
                }
                Some(ObjectType::Blob) => {
                    let path = entry.name().ok_or("invalid path").unwrap();
                    let filepath = Path::new(root).join(path);
                    let filepath = filepath.as_os_str();
                    let blob = object.as_blob().expect("object to be a blob");

                    self.analyze_blob(filepath, blob, results)?;
                }
                _ => continue,
            }
        }
        Ok(())
    }

    fn analyze_blob<P: AsRef<Path>>(
        &self,
        filepath: P,
        blob: &Blob,
        results: &mut IndexMap<PathBuf, Entry>,
    ) -> Result<(), Box<dyn Error>> {
        let contents = blob.content();

        let lang_override = self
            .get_str_attr(&filepath, "gengo-language")?
            .map(|s| s.replace('-', " "))
            .and_then(|s| self.analyzers.get(&s));

        let language =
            lang_override.or_else(|| self.analyzers.pick(&filepath, contents, self.read_limit));

        let language = if let Some(language) = language {
            language.clone()
        } else {
            return Ok(());
        };

        let size = contents.len();
        let generated = self
            .get_boolean_attr(&filepath, "gengo-generated")?
            .unwrap_or_else(|| self.is_generated(&filepath, contents));
        let documentation = self
            .get_boolean_attr(&filepath, "gengo-documentation")?
            .unwrap_or_else(|| self.is_documentation(&filepath, contents));
        let vendored = self
            .get_boolean_attr(&filepath, "gengo-vendored")?
            .unwrap_or_else(|| self.is_vendored(&filepath, contents));

        let detectable = match language.category() {
            Category::Data | Category::Prose => false,
            Category::Programming | Category::Markup | Category::Query => {
                !(generated || documentation || vendored)
            }
        };
        let detectable = self
            .get_boolean_attr(&filepath, "gengo-detectable")?
            .unwrap_or(detectable);

        let path_buf = filepath.as_ref().to_path_buf();
        let entry = Entry {
            language,
            size,
            detectable,
            generated,
            documentation,
            vendored,
        };

        results.insert(path_buf, entry);

        Ok(())
    }

    /// Guesses if a file is generated.
    pub fn is_generated<P: AsRef<Path>>(&self, filepath: P, contents: &[u8]) -> bool {
        self.generated.is_generated(filepath, contents)
    }

    /// Guesses if a file is documentation.
    pub fn is_documentation<P: AsRef<Path>>(&self, filepath: P, contents: &[u8]) -> bool {
        self.documentation.is_documentation(filepath, contents)
    }

    /// Guesses if a file is vendored.
    pub fn is_vendored<P: AsRef<Path>>(&self, filepath: P, contents: &[u8]) -> bool {
        self.vendored.is_vendored(filepath, contents)
    }

    fn get_attr<P: AsRef<Path>>(&self, path: P, attr: &str) -> Result<AttrValue, Box<dyn Error>> {
        let flags = Self::ATTR_CHECK_FLAGS
            .into_iter()
            .reduce(|a, b| a | b)
            .unwrap();
        let attr = self.repository.get_attr(path.as_ref(), attr, flags)?;
        let attr = AttrValue::from_string(attr);
        Ok(attr)
    }

    fn get_boolean_attr<P: AsRef<Path>>(
        &self,
        path: P,
        attr: &str,
    ) -> Result<Option<bool>, Box<dyn Error>> {
        let attr = self.get_attr(path, attr)?;
        let attr = match attr {
            AttrValue::True => Some(true),
            AttrValue::False => Some(false),
            AttrValue::Unspecified => None,
            // NOTE To avoid being overly strict, we'll just ignore invalid attributes.
            _ => None,
        };
        Ok(attr)
    }

    fn get_str_attr<P: AsRef<Path>>(
        &self,
        path: P,
        attr: &str,
    ) -> Result<Option<String>, Box<dyn Error>> {
        let attr = self.get_attr(path, attr)?;
        let attr = match attr {
            AttrValue::String(s) => Some(s),
            AttrValue::Unspecified => None,
            // NOTE To avoid being overly strict, we'll just ignore invalid attributes.
            _ => None,
        };
        Ok(attr.map(String::from))
    }
}

/// A single entry in the language statistics.
#[derive(Debug)]
pub struct Entry {
    /// The detected language.
    language: Language,
    /// The size of the file.
    size: usize,
    /// If the file is detectable (should not be ignored).
    detectable: bool,
    /// If the file was generated.
    generated: bool,
    /// If the file is documentation.
    documentation: bool,
    /// If the file is vendored.
    vendored: bool,
}

impl Entry {
    /// The detected language.
    pub fn language(&self) -> &Language {
        &self.language
    }

    /// The size of the file.
    pub fn size(&self) -> usize {
        self.size
    }

    /// If the file is detectable (should not be ignored).
    pub fn detectable(&self) -> bool {
        self.detectable
    }

    /// If the file was generated.
    pub fn generated(&self) -> bool {
        self.generated
    }

    /// If the file is documentation.
    pub fn documentation(&self) -> bool {
        self.documentation
    }

    /// If the file is vendored.
    pub fn vendored(&self) -> bool {
        self.vendored
    }
}
