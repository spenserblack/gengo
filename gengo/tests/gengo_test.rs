use gengo::Analyzers;
use gengo::Builder;
const ROOT: &str = env!("CARGO_MANIFEST_DIR");

mod util;

#[test]
fn test_javascript() {
    let analyzers = fixture_str!("test_javascript-analyzers.yaml");
    let analyzers = Analyzers::from_yaml(analyzers).unwrap();
    let gengo = Builder::new(ROOT).analyzers(analyzers).build().unwrap();
    let results = gengo.analyze("test/javascript").unwrap();
    insta::assert_debug_snapshot!(results);
}
