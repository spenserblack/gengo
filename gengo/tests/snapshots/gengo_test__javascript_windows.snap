---
source: gengo/tests/gengo_test.rs
expression: results
---
Analysis(
    {
        "bin.js": Entry {
            language: Language {
                name: "JavaScript",
                category: Programming,
                color: "#FFFF00",
            },
            size: 28,
            detectable: true,
            generated: false,
            documentation: false,
            vendored: false,
        },
        "dist\\bin.js": Entry {
            language: Language {
                name: "Plain Text",
                category: Prose,
                color: "#000000",
            },
            size: 62,
            detectable: true,
            generated: true,
            documentation: false,
            vendored: false,
        },
        "docs\\index.html": Entry {
            language: Language {
                name: "HTML",
                category: Markup,
                color: "#FF4400",
            },
            size: 26,
            detectable: false,
            generated: false,
            documentation: true,
            vendored: false,
        },
        "node_modules\\my-dependency\\index.js": Entry {
            language: Language {
                name: "JavaScript",
                category: Programming,
                color: "#FFFF00",
            },
            size: 29,
            detectable: false,
            generated: false,
            documentation: false,
            vendored: true,
        },
        "src\\bin.ts": Entry {
            language: Language {
                name: "TypeScript",
                category: Programming,
                color: "#0000FF",
            },
            size: 62,
            detectable: true,
            generated: false,
            documentation: false,
            vendored: false,
        },
    },
)
